File: MonteCarloTreeSearch/src/main/java/game/IA/UCT.java
package game.IA;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import game.game.Game;
import game.game.Move;

public class UCT {

    public static Move searchForSolution(Game game, Integer timeMax) throws Exception{

        long timeSpentInMillis = 0L;

        Node v0 = createNode(game, null);

        while(timeSpentInMillis < timeMax){
            long begginingTime = System.currentTimeMillis();

            Node v1 = treePolicy(v0);
            Integer delta = defaultPolicy(v1);
            backup(v1, delta);

            timeSpentInMillis+=System.currentTimeMillis()-begginingTime;
        }

        return v0.getMoves().get(bestChild(v0,0));

    }

    private static Node createNode(Game game, Node father) {
        return new Node(game, game.getMoves(), 0, 0, 0, father, new ArrayList<Node>());
    }

    private static Node treePolicy(Node v) throws Exception {

        while (!v.getState().isFinalState()){

            //v.getState().printState();

            if(v.i < v.getMoves().size()){
                System.out.println("Expanding...");
                return expand(v);
            }
                
            else{
                System.out.println("Getting best child...");
                v = v.getChildren().get(bestChild(v, 1/Math.sqrt(2)));
            }
                
        }
        throw new Exception("Abnormal behaviour");
    }

    private static Node expand(Node v) throws Exception {
        Game s = v.getState().applyMovement(v.getMoves().get(v.i));
        v.i = v.i + 1;
        Node child = createNode(s, v);
        v.getChildren().add(child);
        return child;
    }

    private static int bestChild(Node v, double c) {
        Integer iRes = 0;
        Double maxValue = Double.MIN_VALUE;

        for(int i = 0; i<v.getChildren().size(); i++){
            Double value = (v.getChildren().get(i).q/v.getChildren().get(i).n)+c*Math.sqrt(2*Math.log(v.n)/v.getChildren().get(i).n);
            if(value > maxValue){
                maxValue = value;
                iRes = i;
            }
                
        }

        return iRes;
    }

    private static Integer defaultPolicy(Node v) throws Exception {
        Game s = v.getState();
        List<Move> movs = v.getMoves();
        Integer player = v.getFather().getState().getTurn(); 

        while(!s.isFinalState()){
            Move a = movs.get(new Random().nextInt(movs.size()));
            s = s.applyMovement(a);
            movs = s.getMoves();
        }

        if((s.whiteWins() && player == 2) || (s.blackWins() && player == 1))
            return 1;
        else
            return -1;
    }

    private static void backup(Node v, Integer delta) {

        while(v != null){
            v.n = v.n +1;
            v.q = v.q + delta;
            delta = -delta;
            v = v.getFather();
        }

    }
   
    

    


    


    public static void main(String[] args) {

      
    }
    
}

File: MonteCarloTreeSearch/src/main/python/game/UCT.py
from cmath import sqrt
import math
import random
import time
from Game import Game
from UCTNode import UCTNode


class UCT:

    def look_for_solution(s0, t):
        v0 = UCT.create_node(s0, None)
        timePassed = 0
        while(timePassed < t):
            start = time.time()

            v1 = UCT.tree_policy(v0)
            delta = UCT.default_policy(v1)
            UCT.backup(v1, delta)

            end = time.time()
            timePassed += (end - start)
        return v0.moves[UCT.best_child(v0,0)]
            
    
    def create_node(s, father):
        return UCTNode(s, 0, 0, 0, [], father)

    def tree_policy(v):
        while(not v.state.is_final_state()):
            if v.i < len(v.state.get_moves()):
                return UCT.expand(v)
            else:
                v = v.children[UCT.best_child(v, 1/sqrt(2))]
        return v

    def expand(v):
        s = v.state.apply_move(v.moves[v.i])
        v.i = v.i +1 
        child = UCT.create_node(s,v)
        v.children.append(child)
        return child

    def best_child(v,c):
        iRes = 0
        valMax = -99999999.0
        for i in range(0, len(v.children)):
            value = ((v.children[i].q/v.children[i].n) + c*sqrt(2*math.log(v.n)/v.children[i].n)).real
            if value > valMax:
                valMax = value
                iRes = i
        return iRes


    def default_policy(v):
        s = v.state
        movs = v.moves
        player = v.father.state.turn

        while(not s.is_final_state()):
            a = movs[random.randint(0, len(movs)-1)]
            s = s.apply_move(a)
            movs = s.get_moves()

        if (s.white_wins() and player == 2) or (s.black_wins() and player == 1):
            return 1
        else:
            return -1


    def backup(v, delta):
        while v != None:
            v.n = v.n + 1
            v.q = v.q + delta
            delta = -delta
            v = v.father


game = Game.get_initial_state(1, 100)
game.print_state()  

while(not game.is_final_state()):
    game = game.apply_move(UCT.look_for_solution(game, 5))
    game.print_state()
    if game.is_final_state():
        break
    game = game.apply_move(game.get_moves()[0])
    game.print_state()